Anki: Darwin support

matthewbauer opened the pull request:
###### Motivation for this change

These are all changes needed to get Anki working on macOS. No changes to the Anki derivation are necessary. However, to run Anki correctly you must set DYLD_FRAMEWORK_PATH correctly until #20484 is fixed:

```
DYLD_FRAMEWORK_PATH=/System/Library/Frameworks `nix-build '<nixpkgs>' -A anki --no-out-link`/bin/anki
```

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

matthewbauer committed:
mplayer: add darwin platform

matthewbauer committed:
pulseaudio: fix on Darwin

matthewbauer committed:
portaudio: fix on Darwin

matthewbauer committed:
qt4: build webkit on Darwin

QtWebkit is needed by Anki

LnL7 commented on pkgs/servers/pulseaudio/default.nix:
```diff
@@ -41,7 +43,8 @@ stdenv.mkDerivation rec {
   };
 
-  patches = [ ./caps-fix.patch ];
+  patches = [ ./caps-fix.patch ]
```
Where did this patch come from? It looks like you can probably use `fetchpatch` instead.

grahamc replied:
Note this patch was already here

matthewbauer replied:
I think he was referring to the below one.

LnL7 replied:
Indeed, the changes are good now apart from the xcode dependency.

LnL7 commented on pkgs/development/libraries/portaudio/default.nix:
```diff
@@ -12,7 +13,9 @@ stdenv.mkDerivation rec {
+  propagatedBuildInputs = stdenv.lib.optionals stdenv.isDarwin [ AudioUnit AudioToolbox CoreAudio CoreServices Carbon ];
 
   preBuild = stdenv.lib.optionalString stdenv.isDarwin ''
     sed -i '50 i\
```
Not your changes, but this looks really fragile.

matthewbauer replied:
Yep. It's needed here because Makefile is not generated until after configurePhase. We should just modify Makefile.in but the line numbers won't be the same.

LnL7 replied:
I would prefer a patch that fixes the `Makefile.in` then, if that's possible.

matthewbauer committed:
portaudio: fixup patches

- don't use preBuild for patching
- leave all of the tests in place (no tests are run anyway)

LnL7 commented:
This builds fine, but running it seems to require xcode. Is that expected?

matthewbauer commented:
@LnL7 It shouldn't require Xcode. Do you have any clues as to where it uses Xcode?

mdaiter commented on pkgs/applications/video/mplayer/default.nix:
```diff
@@ -199,6 +199,6 @@ stdenv.mkDerivation rec {
     license = "GPL";
     maintainers = [ stdenv.lib.maintainers.eelco stdenv.lib.maintainers.urkud ];
-    platforms = stdenv.lib.platforms.linux;
+    platforms = stdenv.lib.platforms.unix;
```
Hesitations about changing this to "unix" when the title says "Darwin support". Unix is defined to be Linux, Darwin, FreeBSD, OpenBSD, NetBSD and Illumos; as defined here: https://github.com/NixOS/nixpkgs/blob/56904d7c423f2b13b37fbd29f39bbb4b52bc7824/lib/platforms.nix#L21 . This might want to be formatted as `with stdenv.lib.platforms; linux ++ darwin ;` to conform to the header.

LnL7 replied:
I personally don't like being to restrictive with the platforms, IMHO this should include all the platforms that are supported upstream even if tjey doesn't necessarily work (yet) in nixpkgs.

mdaiter commented on pkgs/development/libraries/qt-4.x/4.8/default.nix:
```diff
@@ -4,7 +4,7 @@
 , libtiff, glib, icu, mysql, postgresql, sqlite, perl, coreutils, libXi
 , buildMultimedia ? stdenv.isLinux, alsaLib, gstreamer, gst_plugins_base
-, buildWebkit ? stdenv.isLinux
+, buildWebkit ? true
```
Why is this now set to `true`? Seems like a change for any library dependent on qt-4.8. Why not `buildWebkit ? stdenv.isLinux | stdenv.isDarwin`?

mdaiter reviewed (changes requested):
Just trying to understand some minor changes in here -- looks good, but particularly concerned about the change from `linux` to `unix` when the title suggests only Darwin support.

copumpkin commented on pkgs/development/libraries/portaudio/default.nix:
```diff
@@ -12,30 +13,26 @@ stdenv.mkDerivation rec {
 
   configureFlags = stdenv.lib.optionals stdenv.isDarwin
-    [ "--build=x86_64" "--without-oss" "--enable-static" "--enable-shared" ];
+    [ "--disable-mac-universal" ];
```
Does this flag do any harm when passed into a non-darwin build? I generally try to minimize platform-conditional stuff.

matthewbauer replied:
it will just appear on stdenv..isDarwin

copumpkin replied:
That's what I mean. Does it need to be conditional at all? Wouldn't a linux build just ignore the flag?

matthewbauer replied:
The conditional is probably not needed but it will cause a rebuild without it.

LnL7 replied:
I also prefer not to use isDarwin conditionals if not necessary, is the resulting rebuild significant?

matthewbauer replied:
No, I can change it

LnL7 commented:
https://gist.github.com/9a33c428ab2748fcba50dd5727600bb4

matthewbauer commented:
@LnL7 see 7165b389084966b7a5c96e9b512b7312ae9c676d

LnL7 commented:
Oh, I'll take a closer look at the error then.

vbgl commented:
`gst-plugins-base` does not build: `ld` complains about a missing framework (this is probably unrelated to this PR). After fixing it (patch below), I could build `anki`, but running it results in SIGSEGV.

OS version: Mac OS X 10.11.6 (15G1212)

```
--- a/pkgs/development/libraries/gstreamer/legacy/gst-plugins-base/default.nix
+++ b/pkgs/development/libraries/gstreamer/legacy/gst-plugins-base/default.nix
@@ -1,6 +1,7 @@
 { fetchurl, stdenv, pkgconfig, python, gstreamer, xorg, alsaLib, cdparanoia
 , libogg, libtheora, libvorbis, freetype, pango, liboil, glib, cairo, orc
 , libintlOrEmpty
+, darwin
 , # Whether to build no plugins that have external dependencies
   # (except the ALSA plugin).
   minimalDeps ? false
@@ -36,6 +37,8 @@ stdenv.mkDerivation rec {
         liboil ]
     # can't build cdparanoia on darwin
     ++ stdenv.lib.optional (!minimalDeps && !stdenv.isDarwin) cdparanoia
+    ++ stdenv.lib.optional stdenv.isDarwin
+       (with darwin.apple_sdk.frameworks; ApplicationServices)
     ++ libintlOrEmpty;

   NIX_CFLAGS_COMPILE = stdenv.lib.optionalString stdenv.isDarwin "-lintl";
```

LnL7 commented:
@vbgl Are you on the latest master? http://hydra.nixos.org/build/46408265

vbgl commented:
After updating to master, `gst-plugins-base` indeed builds correctly. But `anki` still crashes before it starts.

matthewbauer commented:
@LnL7 Do you know which derivation requires Xcode to be installed? Most likely it is some bogus version check.

LnL7 commented:
Oh right, for get about that. It's either fixed or I just did something strange when testing.

LnL7 reviewed (approved):
testing, I'll merge once it's finished.
